gk/fib.c

```diff
@@ -0,0 +1,1409 @@
+			__func__, ip_prefix_fib->action);
+		ret = -1;
+		break;
+	}
+
+	/* Reset the fields of the deleted FIB entry. */
+	initialize_fib_entry(ip_prefix_fib);
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"gk: delete a fib entry [ip prefix = %s]\n", ip_prefix->str);
+
+	return ret;
+}
+
+/*
+ * Initialize a gateway FIB entry.
+ * @gateway the gateway address informaiton.
+ * @ip_prefix the IP prefix,
+ * for which the gateway is responsible.
+ */
+static struct gk_fib *
+init_gateway_fib_locked(struct ip_prefix *ip_prefix, enum gk_fib_action action,
+	struct ipaddr *gw_addr , struct gk_config *gk_conf)
+{
+	int ret, fib_id;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+	struct gk_fib *gw_fib, *neigh_fib;
+	struct ether_cache *eth_cache;
+	struct neighbor_hash_table *neigh_ht;
+	struct gatekeeper_if *iface;
+
+	if (gw_addr->proto != ip_prefix->addr.proto) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to initialize a fib entry for gateway, since the gateway and its responsible IP prefix have different IP versions.\n");
+		return NULL;
+	}
+
+	if (action == GK_FWD_GATEWAY_FRONT_NET)
+		iface = &gk_conf->net->front;
+	else if (likely(action == GK_FWD_GATEWAY_BACK_NET))
+		iface = &gk_conf->net->back;
+	else {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to initialize a fib entry for gateway, since it has invalid action %d.\n",
+			action);
+		return NULL;
+	}
+
+	/* Find the neighbor FIB entry for this gateway. */
+	neigh_fib = find_fib_entry_for_neighbor_locked(
+		gw_addr, action, gk_conf);
+	if (neigh_fib == NULL)
+		return NULL;
+
+	/* Find the Ethernet cached header entry for this gateway. */
+	neigh_ht = &neigh_fib->u.neigh;
+	eth_cache = neigh_get_ether_cache_locked(neigh_ht, gw_addr, iface);
+	if (eth_cache == NULL)
+		return NULL;
+
+	/* Find an empty FIB entry for the Gateway. */
+	fib_id = get_empty_fib_id(ip_prefix->addr.proto, gk_conf);
+	if (fib_id < 0)
+		goto put_ether_cache;
+
+	if (ip_prefix->addr.proto == ETHER_TYPE_IPv4)
+		gw_fib = &ltbl->fib_tbl[fib_id];
+	else
+		gw_fib = &ltbl->fib_tbl6[fib_id];
+
+	/* Fills up the Gateway FIB entry for the IP prefix. */
+	gw_fib->action = action;
+	gw_fib->u.gateway.eth_cache = eth_cache;
+
+	ret = lpm_add_route(&ip_prefix->addr, ip_prefix->len, fib_id, ltbl);
+	if (ret < 0)
+		goto init_fib;
+
+	return gw_fib;
+
+init_fib:
+	initialize_fib_entry(gw_fib);
+
+put_ether_cache:
+	ether_cache_put(neigh_fib, action, eth_cache, gw_addr, gk_conf);
+
+	return NULL;
+}
+
+static struct gk_fib *
+init_grantor_fib_locked(
+	struct ip_prefix *prefix, struct ipaddr *gt_addr,
+	struct ipaddr *gw_addr, struct gk_config *gk_conf)
+{
+	int ret, fib_id;
+	struct gk_fib *gt_fib;
+	struct ether_cache *eth_cache;
+	struct neighbor_hash_table *neigh_ht = NULL;
+	struct gatekeeper_if *iface = &gk_conf->net->back;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	/* Find the neighbor FIB entry for this gateway. */
+	struct gk_fib *neigh_fib = find_fib_entry_for_neighbor_locked(
```

Comment:
1. The first test should be if `gt_addr->proto != ip_prefix->addr.proto`.

2. The second test should verify if `gw_addr->proto != ip_prefix->addr.proto` as in `init_gateway_fib_locked()`.

3. It'd be also good to rename variable `prefix` to `ip_prefix` to match `init_gateway_fib_locked()` closely.